src/Microsoft.AspNet.Mvc.ModelBinding/Binders/KeyValuePairModelBinder.cs

```diff
@@ -9,20 +10,34 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
 {
     public sealed class KeyValuePairModelBinder<TKey, TValue> : IModelBinder
     {
-        public async Task<bool> BindModelAsync(ModelBindingContext bindingContext)
+        public async Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
         {
             ModelBindingHelper.ValidateBindingContext(bindingContext,
                                                       typeof(KeyValuePair<TKey, TValue>),
                                                       allowNullModel: true);
 
-            var keyResult = await TryBindStrongModel<TKey>(bindingContext, "key");
-            var valueResult = await TryBindStrongModel<TValue>(bindingContext, "value");
-
+            var keyResult = await TryBindStrongModel<TKey>(bindingContext, "Key");
+            var valueResult = await TryBindStrongModel<TValue>(bindingContext, "Value");
+            var model = bindingContext.ModelMetadata.Model;
+            var isModelSet = false;
             if (keyResult.Success && valueResult.Success)
             {
-                bindingContext.Model = new KeyValuePair<TKey, TValue>(keyResult.Model, valueResult.Model);
+                model = new KeyValuePair<TKey, TValue>(keyResult.Model, valueResult.Model);
+                isModelSet = true;
+            }
+            else if (!keyResult.Success && valueResult.Success)
+            {
+                bindingContext.ModelState.TryAddModelError(keyResult.ModelStateKey,
+                    Resources.KeyValuePair_BothKeyAndValueMustBePresent);
+            }
+            else if (keyResult.Success && !valueResult.Success)
+            {
+                bindingContext.ModelState.TryAddModelError(valueResult.ModelStateKey,
+                    Resources.KeyValuePair_BothKeyAndValueMustBePresent);
             }
```

Comment:
nit: might be simpler to just short circuit return for the error conditions. Up to you